### ISSUE TITLE
❓ How to use `frameProcessorFps`?

### ISSUE THREAD

**Issue by G0retZ** [margelo/react-native-vision-camera#184]
## What

I've  created a frame processor using ImageLabeler as template.
It completes in 200ms per frame
So I tried `frameProcessorFps={2}` and the default one, but anyway Frames got dropped.
Should it throttle frames before FrameProcessor gets into play or am I missing something?

### Logs

```
2021-06-08 12:38:17.886942+0300  Warning: Frame Processor function took 192 ms to execute. This blocks the video queue from recording, optimize your frame processor!
2021-06-08 12:38:17.887228+0300  [native] VisionCamera.captureOutput(_:didDrop:from:): Dropped a Frame. This might indicate that your Frame Processor is doing too much work. Either throttle the frame processor's frame rate, or optimize your frame processor's execution speed. Frame drop reason: FrameWasLate
2021-06-08 12:38:17.888389+0300  [javascript] [native] VisionCamera.captureOutput(_:didDrop:from:): Dropped a Frame. This might indicate that your Frame Processor is doing too much work. Either throttle the frame processor's frame rate, or optimize your frame processor's execution speed. Frame drop reason: FrameWasLate
```

## Reproducable sample

The frame processor itself

```objectiveC
//
//  VisionCameraQRScanner.m
//
//  Created by Sergei Mitrofanov on 07.06.21.
//

#import <VisionCamera/FrameProcessorPlugin.h>
#import <UIKit/UIKit.h>
#import <React/RCTLog.h>
#import <React/RCTUtils.h>

@interface QRCodeFrameProcessorPlugin : NSObject

+ (CIDetector*) detector;

@end

@implementation QRCodeFrameProcessorPlugin

+ (CIDetector*) detector {
  static CIDetector* detector = nil;
  if (detector == nil) {
    NSDictionary *detectorOptions = @{@"CIDetectorAccuracy": @"CIDetectorAccuracyHigh"};
    detector = [CIDetector detectorOfType:CIDetectorTypeQRCode context:nil options:detectorOptions];
  }
  return detector;
}

static inline id scanQRCodes(CMSampleBufferRef buffer, NSArray* arguments) {
  CVImageBufferRef cvImage = CMSampleBufferGetImageBuffer(buffer);
  CIImage *ciImage = [[CIImage alloc] initWithCVPixelBuffer:cvImage];

  NSArray* features = [[QRCodeFrameProcessorPlugin detector] featuresInImage:ciImage];
  NSMutableArray* results = [NSMutableArray arrayWithCapacity:features.count];
  for (CIQRCodeFeature* feature in features) {
    [results addObject:feature.messageString];
  }

  return results;
}

VISION_EXPORT_FRAME_PROCESSOR(scanQRCodes)

@end
```

## Environment

* Platform(s):
  * [X] iOS
  * [ ] Android
* react-native version: 0.64.1
* react-native-vision-camera version: 2.2.0
* react-native-reanimated version: 2.2.0

**Comment by G0retZ** [comment]
I tried to log the workout itself like this but for unknown reason I got no logs at all

```tsx
  const frameProcessor = useFrameProcessor((frame) => {
    'worklet';

    console.log('Boom frame');

    const codes = scanQRCodes(frame);
    console.log('Boom', codes);
    if (codes && codes.length > 0) {
      setScannedData(codes[0]);
    }
  }, []);
```

**Comment by mrousavy** [comment]
Hi!

The frame processor itself looks good to me, at what resolution are you running this on? If you didn't choose a format, chances are that you are receiving 4k frames. Try to get the `frame.width` property to find this out. Ideally you want to resize (downscale) the frame before running AI on it.

As for the logging, I just remembered that I didn't bind `console.log` yet. You can log using `_log`, which prints to the native Xcode console - I'll try to find a workaround to get the logs into your metro console as well!

**Comment by G0retZ** [comment]
Thank you for fast answer!
Gotcha, will try to set a format now.

**Comment by mrousavy** [comment]
Right, try choosing a format with a `videoWidth` of something low like 720 or even lower. I'm also working on a plugin to freely resize buffers right now: https://github.com/mrousavy/vision-camera-resize-plugin

**Comment by mrousavy** [comment]
Also, `frameProcessorFps` is respected as far as I can see. The frame processor still gets executed twice a second (2 FPS), but the Video Queue itself runs at a higher FPS (for video recordings). 

If you don't use video recordings at all, you can set `fps` to something acceptable for the user (e.g. `30` or `60`), run your frame processor at 2 FPS and optimize the execution speed by downsizing the frames (or using a smaller format in the first place). I can also add a functionality to disable the warning that gets printed.

**Comment by G0retZ** [comment]
Oh, I got it.
So it throttles the frames but this doesn't mean frame processor is allowed to execute any longer than 33ms.

I guess this comment made me think that I have all the time between throttled frames to execute.
```tsx
...
* If you're using higher values, always check your Xcode/Android Studio Logs to make sure your frame processors are executing fast enough
* without blocking the video recording queue.
...
```

So should I read it as I will get 33ms for the Frame processor to complete a task or less than that if `frameProcessorFps` set lower?

**Comment by G0retZ** [comment]
Unfortunately even 192*192 took 170 ms to scan the QR Code...

Is it possible to pass a callback or a promise handle into the frame processor arguments just like to a react native module?

**Comment by mrousavy** [comment]
It is allowed to execute longer, that's perfectly fine. 

If your camera is set to 30 FPS, then that means each frame has 33ms to be processed. Remember that the video recorder is also a sort of frame processing.
If you take longer than 33ms, the next frame will be dropped, but that's only a problem if you're recording videos since it looks like there's a small stutter.

If you don't use video recordings, 200ms is a reasonable time for the frame processor to execute - you can run it 5 times a second (5 FPS), which might be enough for your use-case.

**Comment by mrousavy** [comment]
what does that frame processor do? I'm pretty sure there are QR code scanners that are faster than 200ms

**Comment by mrousavy** [comment]
In other words: Frame Processors have exactly `1 / FPS`ms to execute, regardless of what `frameProcessorFps` is set to. `fps` determines how fast they have to execute. `frameProcessorFps` determines how often they get executed if you don't want to run them on every frame.

"have to" execute is a bold statement, since there are use cases where you would be perfectly fine with dropped frames. For example, if you don't use video recordings, frames can be dropped - that's perfectly fine. If you do use video recordings and like 1 out of 20 frames gets dropped, that might be fine for you too, the resulting video might look a bit stuttery, but I haven't experimented with that yet.

> Is it possible to pass a callback or a promise handle into the frame processor arguments just like to a react native module?

I have implemented that functionality, but I didn't test it yet. It should work perfectly fine, just remember that the callbacks have to be worklets:

```js
  const frameProcessor = useFrameProcessor((frame) => {
    'worklet';

    scanQRCodes(frame, (codes) => {
        _log(`QR code plugin completed! Codes: ${codes}`)
    });
  }, []);
```

or dispatch to JS:

```js
  const onQrCodeRead = useCallback((codes) => {
    console.log(`Got QR code: ${codes}`)
    navigation.push("ProductPage", { articleId: codes[0] })
  }, []);
  const frameProcessor = useFrameProcessor((frame) => {
    'worklet';

    scanQRCodes(frame, (codes) => {
        _log(`QR code plugin completed! Codes: ${codes}`)
       runOnJS(onQrCodeRead)(codes)
    });
  }, []);
```

and if you take a look at the docs, the type for this would be a `RCTResponseSenderBlock`:
![Screenshot 2021-06-08 at 13 20 12](https://user-images.githubusercontent.com/15199031/121176224-42e85180-c85c-11eb-9100-297d8fe60e5e.png)

(intended usage: first argument is result, second argument is error)

This means, you can technically copy the buffer, then dispatch a new thread (`DispatchQueue`) and run the QR code scanning algorithm on there. Then it would not drop any frames since the Camera can continue running the video queue while you run your AI on a separate Thread (`DispatchQueue`)

I haven't tested the callback functionality yet, so please let me know if you encounter any issues 👍

**Comment by G0retZ** [comment]
Thank you a lot for your help!
Now this all became much more clear for me.
Will report once will got the satisfying result for my case :-)

**Comment by G0retZ** [comment]
Ok, looks like it doesn't work after all =(
I feel there is something wrong happening.

Tried to use `runOnJS` and every time I get "Don't try to clone an object backed by a different Runtime" and app crashes immediately
Even if I just call a log function like this
```tsx
const logCodes = (codes) => {
 console.log('--- Codes = ', codes);
}

const frameProcessor = useFrameProcessor((frame) => {
    'worklet';

    const scannedCodes = scanQRCodes(frame);
    runOnJS(logCodes)(scannedCodes)
  }, []);
```
And with `useCallback` wrap also the same error.

Struggling for hours with this ='(
What am I missing with this worklet?

```
2021-06-08 18:10:03.286213+0300  FrameProcessorBindings: Converting worklet to Objective-C callback...
Assertion failed: (object->ctx_ == ctx_ && "Don't try to clone an object backed by a different Runtime"), function cloneObject, file /Users/sergeimitrofanov/Projects/emma-app/node_modules/react-native/ReactCommon/jsi/JSCRuntime.cpp, line 611.
```

**Comment by G0retZ** [comment]
Happens in `getValue` here https://github.com/cuvent/react-native-vision-camera/blob/be30d55df5ac6b588108add4023668120f36706c/ios/Frame%20Processor/FrameProcessorRuntimeManager.mm#L151

**Comment by G0retZ** [comment]
For now implemented with a crutch by dispatching scan results via react native emitter module...
Thanks anyway.

**Comment by mrousavy** [comment]
Sorry, was gone yesterday. That's weird, did you manage to get it working with a SharedValue? Because that definitely works. I will also try to investigate the `runOnJS` issue now.

Also, I wonder if this works with Hermes 🤔

**Comment by G0retZ** [comment]
@mrousavy yes, I tried with SharedValue also.
But for some reason despite of being set in the worklet it was not updated in the UI component and wasn't trigger re-rendering...

So I just gave up on both of these =/

**Comment by mrousavy** [comment]
I'll create an example app demonstrating SharedValues soon, but just to confirm - your issue is resolved for now and your frame processor works?

**Comment by mrousavy** [comment]
Btw, I have just released [v2.3.0](https://github.com/cuvent/react-native-vision-camera/releases/tag/v2.3.0) which replaces the `CMSampleBufferRef` with a custom type (`Frame*`) to also provide information about orientation. This means, you need to change the parameter:

```diff
- static inline id somePlugin(CMSampleBufferRef buffer, NSArray* arguments) {
+ static inline id somePlugin(Frame* frame, NSArray* arguments) {
```

and to get the actual `CMSampleBufferRef` you can simply use `frame.buffer`

**Comment by G0retZ** [comment]
For now I'm using a bit ugly workaround:
I created a native iOS module that  dispatches an event with payload to React Native side via `[self sendEventWithName:@"onQRCodesScanned" body:@{@"codes": codes}];`.
From the Frame processor after I got result I dispatch it using the module and listen for it on RN side.

It doesn't look like an elegant and easy to use solution but at least it works.
So far I don't think this solution is acceptable for a public `react-native-vision-camera` plugin.
